opt: support joins in the optbuilder

rytaft opened the pull request:
This commit is one of several needed to build a "memo"
structure from a parsed SQL query. The memo is the primary data
structure in the query optimizer, used for maintaining the forest of
query plan trees that the optimizer considers as part of its search
for the optimal plan.

This commit includes support for building a memo for queries
which include all types of joins (inner, left, right and outer) with
all types of join syntax (`CROSS JOIN`, `ON`, `NATURAL JOIN`, `USING`, etc.).
The test cases have been adapted from the join tests in logictests.

This code is based on code originally written as part of the
opttoy project.

Release note: None

rytaft committed:
opt: support joins in the optbuilder

This commit is one of several needed to build a "memo"
structure from a parsed SQL query. The memo is the primary data
structure in the query optimizer, used for maintaining the forest of
query plan trees that the optimizer considers as part of its search
for the optimal plan.

This commit includes support for building a memo for queries
which include all types of joins (inner, left, right and outer) with
all types of join syntax (CROSS JOIN, ON, NATURAL JOIN, USING, etc.).
The test cases have been adapted from the join tests in logictests.

This code is based on code originally written as part of the
opttoy project.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22724)
<!-- Reviewable:end -->

RaduBerinde commented:
Things are moving fast! :)
Some of the code is from the current planner, right? (I recognize the merged columns comment).  Can you make a short summary about what significant pieces are taken from where and what pieces are new? Just to know what to pay more attention to. Thanks!

rytaft commented:
Sure will do.  Just found a bug so please hold off on looking too carefully.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5P20LfL6ejDDk3AqN1:b-775pun)*
<!-- Sent from Reviewable.io -->

rytaft commented:
I think I've fixed the bug -- ready for review again.

I've added reviewable comments on each function that was partially or mostly copied from existing code.  All of the relevant existing stuff is in `sql/join.go`, `sql/join_predicate.go`, and `sql/data_source.go`. The biggest logic change is that the existing code has a lot of special cases for anonymous sources.  I've tried to achieve the same semantics with the checks in `findColumn` in `scope.go`, but I'm not 100% sure that they are correct. (`findColumn` will also change again once #22371 is merged...)

---

Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/builder.go, line 266 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PfA3YYYmY8WsK5gcm:-L5PfA3YYYmY8WsK5gcn:bcritgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L266)):*
> ```Go
> 
> // renameSource applies an AS clause to the columns in scope.
> func (b *Builder) renameSource(as tree.AliasClause, scope *scope) {
> ```

mostly copied from `renameSource` in `sql/data_source.go`

---

*[pkg/sql/opt/optbuilder/builder.go, line 911 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PfdxXVTdZN8NPIgZB:-L5PfdxXVTdZN8NPIgZC:bjp7qcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L911)):*
> ```Go
> // See Builder.buildStmt above for a description of the remaining input and
> // return values.
> func (b *Builder) buildJoin(
> ```

partially copied from `makeJoin` and `makeJoinPredicate` in `sql/join.go`

---

*[pkg/sql/opt/optbuilder/builder.go, line 974 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5Pg4gI0d2Ckp_zMunH:-L5Pg4gI0d2Ckp_zMunI:bmg7ff5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L974)):*
> ```Go
> // commonColumns returns the names of columns common on the
> // left and right sides, for use by NATURAL JOIN.
> func commonColumns(leftScope, rightScope *scope) (common tree.NameList) {
> ```

mostly copied from `commonColumns` in `sql/join.go`

---

*[pkg/sql/opt/optbuilder/builder.go, line 1106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PgKz3SE3sZV6EOriq:-L5PgKz3SE3sZV6EOrir:bx0vby5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L1106)):*
> ```Go
> // See Builder.buildStmt above for a description of the remaining input and
> // return values.
> func (b *Builder) buildUsingJoinPredicate(
> ```

partially copied from `makeJoin` in `sql/join.go` (including comment above), and `makeUsingColumns` and `makePredicate` in `sql/join_predicate.go`

---

*[pkg/sql/opt/optbuilder/builder.go, line 1178 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PhQqTd0P4hyGcmDHL:-L5PhQqTd0P4hyGcmDHM:b-qc8w9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L1178)):*
> ```Go
> //     not equal, it is marked as hidden and added to the scope.
> // (3) All other columns are added to the scope without modification.
> func hideMatchingColumns(
> ```

mostly copied from `makeJoin` in `sql/join.go`

---

*[pkg/sql/opt/optbuilder/builder.go, line 1216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PhqDSjkrmwVmVrK5e:-L5PhqDTMtGJADNmqqlA:b-wxslt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L1216)):*
> ```Go
> )
> 
> func getJoinType(astJoinType string) joinType {
> ```

copied from `makeJoin` in `sql/join.go`

---

*[pkg/sql/opt/optbuilder/builder.go, line 1251 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5Pi0Ba3AeZ_yd98OXA:-L5Pi0Ba3AeZ_yd98OXB:b2vv1v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/builder.go#L1251)):*
> ```Go
> // context is a string ("left" or "right") used to indicate in the error
> // message whether the name is missing from the left or right side of the join.
> func findUsingColumn(cols []columnProps, name optbase.ColumnName, context string) *columnProps {
> ```

mostly copied from `pickUsingColumn` in `sql/join_predicate.go`

---

*[pkg/sql/opt/optbuilder/scope.go, line 268 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PiPtyZq-Zw9WgBq3i:-L5PiPtyZq-Zw9WgBq3j:b79b0t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/scope.go#L268)):*
> ```Go
> // searches the parent scope. If the column is not found in any of the
> // ancestor scopes, findColumn throws an error.
> func (s *scope) findColumn(c *tree.ColumnItem) *columnProps {
> ```

partially copied from `findColumn` in `sql/data_source.go`

---

*[pkg/sql/opt/optbuilder/scope.go, line 439 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PidajljUKfi7U-oFH:-L5PidajljUKfi7U-oFI:b-oiiwn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/scope.go#L439)):*
> ```Go
> // ambiguousError returns an error with a helpful error message to be used in
> // case of an ambiguous column reference.
> func ambiguousError(c *tree.ColumnItem, candidates []*columnProps) error {
> ```

copied from `findColHelper` in `sql/data_source.go`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5PlQrsB8KvzE7WkQXB:byx4c74)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks! Do you plan to wait for #22371 to go in before merging this? (that would be good if merging this first would require more work in that PR, since that one is for 2.0)

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, looks good - I like that you're actively pulling in code from existing planner.

A couple of general comments/questions:

1. Have you consistently been matching error messages in existing code? I can't think of a reason why we should create any new error messages unless/until we actually add new functionality that the existing code does not have. If you haven't already, maybe do an audit sweep of the builder code to make sure all messages are the same as 2.0.

2. `builder.go` is getting large and unwieldy. It might be good to begin breaking it up into separate files. I could imagine files like `join.go`, `scalar.go`, `data_source.go`, etc., naming each thing that we're building in that file.

I usually think about refactoring when:
1. Files exceed ~1000 lines.
2. Functions exceed ~100 lines.
3. Directories exceed ~25 non-test files.

There are plenty of good exceptions to those rules of thumb (like auto-generated files, functions with big comment blocks, files that are obviously closely-related, etc). However, if you're exceeding those limits, it's a good idea to review the situation and look for opportunities to recategorize into easier-to-digest units.

---

Review status: 0 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/scope.go, line 294 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5UDlc3A7gtUq0BgLDs:-L5UDlc3A7gtUq0BgLDt:b-g0mv17) ([raw file](https://github.com/cockroachdb/cockroach/blob/19ae95b8782746efba46af2b1cc648b2971534ed/pkg/sql/opt/optbuilder/scope.go#L294)):*
> ```Go
> 	allowHidden := true
> 
> 	for curr := s; curr != nil; curr, allowHidden = curr.parent, false {
> ```

Any opportunities to simplify this block with a helper method or two?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5UE1DMammpVhKs3Qor:b-werpre)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Great work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>   Will take another look when you integrate with Raphael's change (the PR is merged).

Thanks for the annotations, they helped a lot!

---

Review status: 0 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/builder.go, line 1207 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5WOhN8ZOki0aFn83W6:-L5WOhN8ZOki0aFn83W7:b-o9vk0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/19ae95b8782746efba46af2b1cc648b2971534ed/pkg/sql/opt/optbuilder/builder.go#L1207)):*
> ```Go
> }
> 
> type joinType int
> ```

[nit] rebase and use `sqlbase.JoinType` (there is a function to get one from an `astJoinType` too)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5WPahJwZsB6gKB4qS8:bnycgyu)*
<!-- Sent from Reviewable.io -->

rytaft committed:
opt: fix name resolution in the optbuilder

This commit uses the new name resolution code from #22753
to properly resolve names in the optbuilder.

Release note: None

rytaft committed:
opt: use sqlbase.JoinType in optbuilder

Remove the redundant joinType enum from optbuilder
and use the new sqlbase.JoinType enum instead.

Release note: None

rytaft commented:
Thanks for the reviews! I've integrated the new name resolution stuff and join stuff (in separate commits for ease of review).

@andy-kimball, in response to your two questions/comments:
1. I've been careful in the last couple of commits to copy the error messages used by the existing planner. I'll look for other error messages that don't match and fix those in a separate PR.
2. I agree that the builder is getting too big. I'll do the refactor in a separate PR as well to avoid cluttering this one.

---

Review status: 0 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/opt/optbuilder/builder.go, line 1207 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5WOhN8ZOki0aFn83W6:-L5pqbQi0ibFO-Z0q9Qn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19ae95b8782746efba46af2b1cc648b2971534ed/pkg/sql/opt/optbuilder/builder.go#L1207)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] rebase and use `sqlbase.JoinType` (there is a function to get one from an `astJoinType` too)
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/scope.go, line 294 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5UDlc3A7gtUq0BgLDs:-L5pl7x_kYWhsE55vFUa:b-xm2ekh) ([raw file](https://github.com/cockroachdb/cockroach/blob/19ae95b8782746efba46af2b1cc648b2971534ed/pkg/sql/opt/optbuilder/scope.go#L294)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Any opportunities to simplify this block with a helper method or two?
</blockquote></details>

I added a helper method. It doesn't reduce the size of the code, but I think it's still an improvement. (Or did you have something else in mind?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5prkH2_coL1YWLa6C_:bx9z897)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 13 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5tBrTISWSa0eQYhi-d:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
TFTR! @knz could you perhaps take a look as well? No need to review the whole PR, just the stuff related to name resolution (the second commit).

There's also a TODO left over from your commit in `scope.FindSourceProvidingColumn`, and I'm not sure if I've sufficiently addressed it.

---

Review status: 0 of 13 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5tj__sV8Ivq5CEYcx8:b-gyncw6)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM

knz commented:
Review status: 0 of 13 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/scope.go, line 268 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PiPtyZq-Zw9WgBq3i:-L5ulfvXAXtOJ7yUW5t2:b-qvlovb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/scope.go#L268)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

partially copied from `findColumn` in `sql/data_source.go`
</blockquote></details>

This `findColumn` code was obsolete and was entirely replaced. Check out `sqlbase/column_resolver.go` and if possible reuse as much of it as possible.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5ukvF78yEnDIK6vFGe:b64e94n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
> ```Go
> 
> // FindSourceProvidingColumn is part of the tree.ColumnItemResolver interface.
> func (s *scope) FindSourceProvidingColumn(
> ```

I think the logic here is incomplete.

`SELECT k FROM kv a, kv b`is ambiguous (your code does this properly)

however

`SELECT k FROM kv a, kv b, (SELECT 1 as k)`

is not

i.e. there is no ambiguity if the requested column name is unqualified and there is exactly one anonymous source that provides that column name.

see `FindSourceProvidingColumn` in `column_resolver.go`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the reviews!

---

Review status: 0 of 13 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/scope.go, line 268 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5PiPtyZq-Zw9WgBq3i:-L5y0C9mCBedEiG4_Bde:b27596o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa79904be84dd3f47d3e6f5db28412bdb430ba3/pkg/sql/opt/optbuilder/scope.go#L268)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This `findColumn` code was obsolete and was entirely replaced. Check out `sqlbase/column_resolver.go` and if possible reuse as much of it as possible.
</blockquote></details>

Yep I checked that code.  My logic looks slightly different since we aren't storing the sources separately from the columns, but it should be semantically equivalent.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5xzfXTXfn6gYmM77yj:bkbiz5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think the logic here is incomplete.

`SELECT k FROM kv a, kv b`is ambiguous (your code does this properly)

however

`SELECT k FROM kv a, kv b, (SELECT 1 as k)`

is not

i.e. there is no ambiguity if the requested column name is unqualified and there is exactly one anonymous source that provides that column name.

see `FindSourceProvidingColumn` in `column_resolver.go`
</blockquote></details>

That case is handled in the default case of the switch statement below (see line 337).  There is also a test case on line 43 of `opt/optbuilder/testdata/join` which is similar to your example - I added a reviewable comment so you can find it.

---

*[pkg/sql/opt/optbuilder/testdata/join, line 43 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5y-Xep5CSX_Ap6TkLz:-L5y-Xep5CSX_Ap6TkM-:b-8akbl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/testdata/join#L43)):*
> ```
> # that also provide it.
> build
> SELECT x FROM (SELECT 1 AS x), onecolumn AS a, onecolumn AS b
> ```

Here is the test case for a single anonymous source.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5y0eYQVDc6h0DoTD4C:b-hlc8bm)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 13 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5y4FXemfLCJbrTi0Gl:b-q024ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

That case is handled in the default case of the switch statement below (see line 337).  There is also a test case on line 43 of `opt/optbuilder/testdata/join` which is similar to your example - I added a reviewable comment so you can find it.
</blockquote></details>

Ahah but then it's not correct either:

`SELECT k FROM kv, (SELECT 1 AS k)` should return the 2nd k, not the first one. 

Your test didn't catch this because 1) it's only using ambiguous named soruces 2) the anonymous source is last.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 13 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5y5h9IdLXmJf8wgHLW:b52evuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ahah but then it's not correct either:

`SELECT k FROM kv, (SELECT 1 AS k)` should return the 2nd k, not the first one. 

Your test didn't catch this because 1) it's only using ambiguous named soruces 2) the anonymous source is last.
</blockquote></details>

It does actually handle that correctly - I've added a test case using your example. You can see from the test output that it's choosing the second k, not the first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 13 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5y75gE729gfgv1IDj2:btknbky) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It does actually handle that correctly - I've added a test case using your example. You can see from the test output that it's choosing the second k, not the first.
</blockquote></details>

Can you walk me through the code to understand the path that makes this possible? I don't see it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 13 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5yBdrF4kGCAIrjTBM0:b-257tju) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you walk me through the code to understand the path that makes this possible? I don't see it.
</blockquote></details>

So after talking face to face, I am convinced the code is correct. Thanks for the explanation.

Just wanted to propose to get rid of the candidate slice allocation and make the control flow clearer:

```
var candidateWithPrefix *columnProps
var candidateFromAnonymousSource *columnProps
moreThanOneCandidateWithPrefix := false
moreThanOneCandidateFromAnonSource := false

...

if col.name == colName && (allowHidden || !col.hidden) {
	if col.table.TableName == "" {
               if candidateFromAnonSource != nil { moreThanOneCandidateFromAnonSource = true }
               candidateFromAnonSource = col
        } else {
               if candidateWithPrefix != nil { moreThanOneCandidateWithPrefix = true }
               candidateWithPrefix = col
        }
}

...

if candidateFromAnonSource != nil && !moreThanOneCandidateFromAnonSource {
    return candidateFromAnonSource
}
if moreThanOneCandidateFromAnonSource || moreThanOneCandidateWithPrefix {
   // error
}
return candidateWithPrefix
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 13 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/scope.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5ukvF78yEnDIK6vFGd:-L5y7W2NzRUQNDPg1ngJ:b-hwtxj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bfea77b4e69bb46cd8ac167f9f5bc59175f1228/pkg/sql/opt/optbuilder/scope.go#L278)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

So after talking face to face, I am convinced the code is correct. Thanks for the explanation.

Just wanted to propose to get rid of the candidate slice allocation and make the control flow clearer:

```
var candidateWithPrefix *columnProps
var candidateFromAnonymousSource *columnProps
moreThanOneCandidateWithPrefix := false
moreThanOneCandidateFromAnonSource := false

...

if col.name == colName && (allowHidden || !col.hidden) {
	if col.table.TableName == "" {
               if candidateFromAnonSource != nil { moreThanOneCandidateFromAnonSource = true }
               candidateFromAnonSource = col
        } else {
               if candidateWithPrefix != nil { moreThanOneCandidateWithPrefix = true }
               candidateWithPrefix = col
        }
}

...

if candidateFromAnonSource != nil && !moreThanOneCandidateFromAnonSource {
    return candidateFromAnonSource
}
if moreThanOneCandidateFromAnonSource || moreThanOneCandidateWithPrefix {
   // error
}
return candidateWithPrefix
```
</blockquote></details>

Ok I think I finally have something that works - thanks for the suggestion!  I had to modify the code above a bit to account for the special case of hidden columns, but otherwise it's pretty much the same.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

rytaft committed:
opt: add typing for coalesce expression

Add a function to infer the type of a coalesce expression
based on the types of its children.

Release note: None

rytaft commented:
I had to add type inference for the `CoalesceOp` operator after rebasing to avoid errors. Please take another look.

---

Review status: 0 of 14 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5zAbDlPbjDEFeGXWs_:b17yfm1)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 15 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/scope.go, line 334 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5zSmnGvK9mYJTHX95I:-L5zSmnH_m4LScwGQdup:b-9jh3ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e4f94e7965b95830c8a498807f5ae08c09ed17/pkg/sql/opt/optbuilder/scope.go#L334)):*
> ```Go
> 					hiddenCandidate = col
> 				}
> 			}
> ```

I just realized this loop is still missing a conditional. Consider this:

`select * from (select 1 as k), (select 2 as k) where 1 in (select k from kv)`

I think your code currently rejects the subquery on the right saying the name `k` is ambiguous for the anonymous source, because it will loop over the scopes and discover the 2 k's defined by the subqueries on the left.

But that's not correct: `k` can be resolved unambiguously in the nested subquery on the right.

The problem is that the conditional about ambiguity (below this comment) must be tested at each level of scoping, not after you've tested all scopes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TypeCoalesce LGTM

rytaft commented:
Review status: 0 of 15 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/scope.go, line 334 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5zSmnGvK9mYJTHX95I:-L5zY2QpCer4tOSdKqZP:b-jjeqyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e4f94e7965b95830c8a498807f5ae08c09ed17/pkg/sql/opt/optbuilder/scope.go#L334)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I just realized this loop is still missing a conditional. Consider this:

`select * from (select 1 as k), (select 2 as k) where 1 in (select k from kv)`

I think your code currently rejects the subquery on the right saying the name `k` is ambiguous for the anonymous source, because it will loop over the scopes and discover the 2 k's defined by the subqueries on the left.

But that's not correct: `k` can be resolved unambiguously in the nested subquery on the right.

The problem is that the conditional about ambiguity (below this comment) must be tested at each level of scoping, not after you've tested all scopes.

</blockquote></details>

Good catch!  I think the issue is that I haven't yet implemented subquery replacement.  I just did a local test on my laptop by adding in the subquery replacement code from opttoy, and then this case works correctly.  Without subquery replacement it fails for the reason you identified.  

Perhaps I can save this particular case for the next PR with subquery replacement?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 15 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/scope.go, line 334 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5zSmnGvK9mYJTHX95I:-L5z_7bcw2lbwsGuWzsO:brhvj6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e4f94e7965b95830c8a498807f5ae08c09ed17/pkg/sql/opt/optbuilder/scope.go#L334)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Good catch!  I think the issue is that I haven't yet implemented subquery replacement.  I just did a local test on my laptop by adding in the subquery replacement code from opttoy, and then this case works correctly.  Without subquery replacement it fails for the reason you identified.  

Perhaps I can save this particular case for the next PR with subquery replacement?
</blockquote></details>

Then at least prepare the test cases (albeit potentially disabled) so we have something to remind ourselves this needs to be improved

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks @RaduBerinde and @knz!

---

Review status: 0 of 15 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/opt/optbuilder/scope.go, line 334 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-L5zSmnGvK9mYJTHX95I:-L5zaAsYVFgDl6siiKMM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e4f94e7965b95830c8a498807f5ae08c09ed17/pkg/sql/opt/optbuilder/scope.go#L334)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Then at least prepare the test cases (albeit potentially disabled) so we have something to remind ourselves this needs to be improved
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22724#-:-L5z_FD21nQ07WeI41jY:bo34oi0)*
<!-- Sent from Reviewable.io -->
